### ISSUE TITLE
Perhaps the face detector is due for an upgrade

### ISSUE THREAD

**Issue by TalosOfCrete** [iperov/DeepFaceLab#760]
Seeing as S3FD is beginning to show its age - that being three years old - due to many users' difficulty using it on more complex scenes, such as ones with poor lighting and/or an abundance of occlusions, I believe has come time to retire S3FD and replace it with one of the multitude of face alignment papers that are far more recent and robust; these newer face alignment algorithms can be found [here](https://paperswithcode.com/task/face-alignment/latest#code), and seem to be substantially more resistant to occlusions and poor lighting.  The only concerns I would have with this upgrade would primarily be how ingrained S3FD is into this repository, i.e. if it can merely be swapped with a pretrained model and some variable changes, or if it would require a more fundamental restructuring of the codebase.

I am opening this as an issue to promote discussion relating to the current face extraction algorithm and whether it is indeed sufficient.  I, however, am still planning to upgrade the detector on my fork of DFL [here](https://github.com/TalosOfCrete/DeepFaceLab).  Please leave your thoughts in this issue thread.  I would especially love to hear @iperov 's take on this.

**Comment by iperov** [comment]
S3FD just searches the face rectangle. 
But the rectangle can be placed manually. 
My version of landmarks detector is better than original 2DFAN : first it has 2 passes, second it uses super sampling. So it slower but more accurate.
But face landmarks are no more important with WF+XSeg, because the masks provided by user.
WF+XSeg is currently the best solution to create high quality deepfake.

Therefore face landmarks are used only for face alignment.
But strict face alignment is not needed, because samples are randomly rotated and displaced during the training.

**Comment by iperov** [comment]
also S3FD is third face detector.
First was DLIB-CNN 
then MTCNN
and now S3FD

**Comment by TalosOfCrete** [comment]
Understood.  I am currently training my own RetinaFace model, and plan on plugging it into my fork of DFL.  Once I verify that it is functional, I may initiate a pull request if it proves to be more robust.

**Comment by TalosOfCrete** [comment]
I have also been searching for more recent landmark detection models, and have found ones that are quite interesting.

FYI: My current DFL fork is experimenting with depthwise separable convolutions in place of the standard convolutions in an effort to reduce parameter count at a minimal-to-no quality loss.  However, I and Ctrl+Shift+Face are currently testing it for perceptible quality drop.  In terms of parameter efficiency, however, it seems to approximately halve the parameter counts when placed in the encoder and decoder.  If there is no quality loss, I will initiate a pull request for that too.  If there is, I will implement it as a model option and pull request that.

My only concern is that S3FD often fails to properly detect faces in more challenging scenes, and that even your more robust landmark detector can - at times - struggle with odd occlusions/shadows.  I have found papers using even more novel techniques than the one you had initially used.

Tell me, though, because Ctrl was complaining about this during his most recent fake: since you said the rectangle was generated by S3FD, then I assume that manual rectangle selection is a substitute for that, correct?  If that is indeed the case, I have found that the manual pass, while quite robust, has difficulty detecting the landmarks on more challenging spots.  Perhaps both the automatic detection algorithm and - perhaps more importantly - the landmark detector could be improved?  I recognize your own improvements upon FAN, but it is still a two year old paper, whereas recent advances in landmark detection have shown exceptionally impressive results (such as [this](https://arxiv.org/pdf/1904.07399v3.pdf), [this](https://jmbuena.github.io/publications/PRL_2019_web.pdf), and [this](https://github.com/emilianavt/OpenSeeFace); the last one merely implements and improves upon the first paper mentioned) @iperov .

**Comment by TalosOfCrete** [comment]
@iperov I currently have a prototype that uses RetinaFace for the face rectangles and Adaptive Wing Loss FAN for the landmarks.  However, while it functions, it does so in the "does not throw exceptions" way.  Put it another way: it runs, but it won't properly find the rectangles for RetinaFace (which is odd; essentially what happens is that it detects over 3000 bounding boxes with low confidence, but last night I was able to get it to detect only two bounding boxes with high confidence) and appears to be centering incorrectly for the landmarks.  Perhaps you could take a look at my fork?  I may initiate a pull request for this.

Do bear in mind that I could not find any native TensorFlow implementations of these, nor could I successfully convert them using ONNX or MMDNN, so they are running natively in PyTorch (this has been added to the requirements.txt, along with a PyTorch-based implementation of nn.py).

**Comment by iperov** [comment]
> Perhaps you could take a look at my fork? 

no

**Comment by TalosOfCrete** [comment]
@iperov Why not?  These algorithms have demonstrably worked, and are empirically superior to the current ones (their papers being found [here](https://arxiv.org/abs/1905.00641) and [here](https://arxiv.org/abs/1904.07399), with their results and examples pictured below (figures 1 and 2 for RetinaFace; figures 3, 4, and 5 for AWL).  In my view, it is indeed worth a look into making it functional in DFL, as it would expand the range of scenes the repository is capable of processing, allowing it to further outstrip its competition from FaceSwap.  In fact, you do mention RetinaFace [in your own paper](https://arxiv.org/abs/2005.05535) (figure 6).

I implore you to examine this  (#767) , as you are substantially more familiar with the face extraction code than I.  I believe that these two alone could push DFL so far ahead of the competition that it will truly become the premier repository for creating deepfakes.

![Screen Shot 2020-06-02 at 12 47 51 AM](https://user-images.githubusercontent.com/32628941/83485012-ba2a1f00-a46b-11ea-9c41-32f03d7c3957.png)
**Figure 1:** Results of RetinaFace on WIDER Face Val: Hard subset (S3FD is highlighted)

![Screen Shot 2020-06-02 at 12 48 20 AM](https://user-images.githubusercontent.com/32628941/83485132-fe1d2400-a46b-11ea-8993-978ee1b61d0a.png)
**Figure 2:** Results of RetinaFace on WIDER Face Test: Hard subset (S3FD is highlighted)

![Screen Shot 2020-06-02 at 12 52 27 AM](https://user-images.githubusercontent.com/32628941/83485277-548a6280-a46c-11ea-8b1d-7734cea7fcbb.png)
**Figure 3** AWL evaluation on WFLW

![Screen Shot 2020-06-02 at 12 52 37 AM](https://user-images.githubusercontent.com/32628941/83485320-679d3280-a46c-11ea-8b1f-872161cf25a8.png)
**Figure 4:** AWL evaluation on 300W

![wflw](https://user-images.githubusercontent.com/32628941/83485366-800d4d00-a46c-11ea-9603-4c095704fce2.png)
**Figure 5:** Selected samples from AWL paper.  Note the exceptional landmark alignment in even the worst of conditions.

![Screen Shot 2020-06-02 at 12 54 32 AM](https://user-images.githubusercontent.com/32628941/83485426-adf29180-a46c-11ea-8756-128d3d1db267.png)
**Figure 6:** An excerpt from your own paper describing alternatives to S3FD

**Comment by iperov** [comment]
1) your PR is shit and breaking DFL
2) I will never use pytorch.
3)
![firefox_2020-06-02_10-31-48](https://user-images.githubusercontent.com/8076202/83487544-4401d880-a4bc-11ea-9ddd-31f53fe6a3ca.jpg)

such landmarks are unacceptable. Side face must have the jaw line on the cheek line in order to properly fill the mask on the mask.

**Comment by TalosOfCrete** [comment]
@iperov First, that’s rather blunt.  Second, the frames I showed (and that you showcased as an awful alignment) were ones that were completely missed by S3FD and FAN, and were exceedingly difficult to get on manual.  In fact, given the original frame, this is actually a perfect landmark alignment.

Further, that’s why I added the PR as a draft.  It’s currently a work in progress, but it will soon be integrated properly into DFL, at which point I will do two things:
1. Turn it into an actual PR and request reviewers
2. Release it to major deepfake YouTubers on CSF’s Discord, who may lobby for its addition on my behalf.  You do realize that CSF’s most recent fake would’ve been impossible with just DFL’s normal tools (and the conclusions were merely cracks in a windshield) and the landmarks had to be manually edited with our own, home-brewed software.

Lastly, I see no reason to not use PyTorch in this case.  The models and scripts are already written as such, making implementation easy; further, it boasts no conflicts with DFL’s TensorFlow.

**Comment by iperov** [comment]
I only see one way out.
If you want retinaface detector (not landmark extractor) in DFL.
Port it to TF with weights file, compare output of pytorch and tf version. Diff should be < 0.001
provide minimal usage example. I will check it and integrate as second detector.

**Comment by iperov** [comment]
also don't dislike my posts, if you don't want to banned.

**Comment by TalosOfCrete** [comment]
Banning people for disliking posts?  Quite harsh.  I just found the reply to be cursory, brash, and uncivil, and believe it warranted the dislike.

Second, I did attempt to port the files to TF.  Two different avenues (ONNX and MMDNN) were unsuccessful, as both use advanced custom functions unsupported by those conversion methods.

As it stands right now, there are no PyTorch-TF conflicts whatsoever in the code.  It only needs some finishing touches to be properly integrated.  Perhaps you could enumerate your reasons for your insistence on TF-only - as I have found that mixing neural network frameworks very often encounters no problems - and I may respond to them in kind.

Besides, using the natively trained PyTorch versions ensures consistency in results.  It is the premier framework for researchers, hence the predominant implementations of these algorithms are in PyTorch.  If they are in TF, they quite often don’t include pre-trained models.  Training f one of these myself would require I redactor the code into TF and then follow their training regimen perfectly, while simply using the PyTorch versions allows for the performance of the model to be delivered as advertised.

**Comment by TalosOfCrete** [comment]
It’s all in Facelib.  Again, it currently doesn’t properly find the bounding boxes within DFL’s framework, so I was hoping someone with much more knowledge about DFL than I could help debug it.

**Comment by iperov** [comment]
so RetinaFaceExtractor does not work properly?
